substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,581 @@
+            }
+            return sb;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            InvocationResultInline that = (InvocationResultInline) o;
+            return site.equals(that.site) &&
+                            Objects.equals(callee, that.callee) &&
+                            children.equals(that.children);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(site, callee, children);
+        }
+    }
+}
+
+/**
+ * This class detects what can be inlined.
+ */
+class TrivialMethodDetector {
+
+    private final BigBang bb;
+    private final HostedProviders providers;
+    private final GraphBuilderConfiguration prototypeGraphBuilderConfig;
+    private final OptionValues options;
+    private final DebugContext debug;
+
+    TrivialMethodDetector(BigBang bb, HostedProviders providers, GraphBuilderConfiguration originalGraphBuilderConfig, OptionValues options, DebugContext debug) {
+        this.bb = bb;
+        this.debug = debug;
+        this.prototypeGraphBuilderConfig = makePrototypeGraphBuilderConfig(originalGraphBuilderConfig);
+        this.options = options;
+        this.providers = providers;
+    }
+
+    private static GraphBuilderConfiguration makePrototypeGraphBuilderConfig(GraphBuilderConfiguration originalGraphBuilderConfig) {
+        GraphBuilderConfiguration result = originalGraphBuilderConfig.copy();
+
+        result.getPlugins().clearInlineInvokePlugins();
+        for (InlineInvokePlugin inlineInvokePlugin : originalGraphBuilderConfig.getPlugins().getInlineInvokePlugins()) {
+            if (!(inlineInvokePlugin instanceof NativeImageInlineDuringParsingPlugin)) {
+                result.getPlugins().appendInlineInvokePlugin(inlineInvokePlugin);
+            }
+        }
+
+        return result;
+    }
+
+    @SuppressWarnings("try")
+    InvocationResult analyzeMethod(CallSite callSite, AnalysisMethod method, ValueNode[] args) {
+        if (!method.hasBytecodes()) {
+            /* Native method. */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        } else if (providers.getGraphBuilderPlugins().getInvocationPlugins().lookupInvocation(method) != null) {
+            /* Method has an invocation plugin that we must not miss. */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        } else if (method.isSynchronized()) {
+            /*
+             * Synchronization operations will always bring us above the node limit, so no point in
+             * starting an analysis.
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        }
+
+        MethodNodeTrackingAndInline methodState = new MethodNodeTrackingAndInline(new InvocationResultInline(callSite, method));
```

Comment:
`MethodNodeTrackingAndInline.result` is not actually used in the class, it is just stored. Why not just allocate the `InvocationResultInline` below, in the `return` statement?

Reply:
Actually, we use it to store decisions about children inlining.